controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -186,29 +208,37 @@ private void triggerReadyJobs(LockedApplication application) {
      */
     private boolean changesAvailable(Application application, JobStatus previous, JobStatus next) {
         if ( ! application.deploying().isPresent()) return false;
-        Change change = application.deploying().get();
 
-        if ( ! previous.lastSuccess().isPresent()) return false;
+        Change change = application.deploying().get();
 
-        if (change instanceof Change.VersionChange) {
+        if (change instanceof Change.VersionChange) { // Propagate upgrade while making sure we never downgrade
             Version targetVersion = ((Change.VersionChange)change).version();
-            if ( ! (targetVersion.equals(previous.lastSuccess().get().version())) )
-                return false; // version is outdated
-            // The below is checked again in allowedTriggering, right before actual triggering.
-            if (next != null && isOnNewerVersionInProductionThan(targetVersion, application, next.type()))
-                return false; // Don't downgrade
-        }
 
-        if (next == null) return true;
-        if ( ! next.lastSuccess().isPresent()) return true;
+            // Is the target version tested?
+            if ( ! lastSuccessfulIs(targetVersion, JobType.stagingTest, application))
+                return false;
```

Comment:
I can't get this to be right:
1. System test succeeds with version _A_.
2. Staging test dies with version _A_, i.e., no report is generated, and the job is left hanging, to be triggered by the ReadyJobTrigger.
3. We now enter this method, looking to see if we should trigger `stagingTest` with the target version _A_, but then  `lastSuccessfulIs(A, stagingTest, app)` is `false`, so we return `false`, and don't trigger `stagingTest`, which we should really do. 

Am I just wrong?

Reply:
Ah, `isHanging` to the rescue again, huh?

Reply:
I think you are right (about the last thing). Fixed.